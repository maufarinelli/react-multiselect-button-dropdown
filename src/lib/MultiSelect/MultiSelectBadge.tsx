import React from 'react';
import styled from 'styled-components';

// TODO: theming for colors
const Badge = styled('span')`
  padding: 0.1875rem 0.5rem; /* 3px 8px if base font-size is 16px */
  margin-left: 0.3125rem;
  font-size: 0.85rem;
  font-weight: normal;
  line-height: 1;
  color: #fff;
  vertical-align: middle;
  background-color: #000;
  border-radius: 50%;
`;

interface MultiSelectBadgeProps {
  text: string;
  quantity: number;
}

const MultiSelectBadge: React.FC<MultiSelectBadgeProps> = ({ text, quantity }) => (
  <>
    {text} <Badge className="multiselect-badge">{quantity}</Badge>
  </>
);
export default MultiSelectBadge;
