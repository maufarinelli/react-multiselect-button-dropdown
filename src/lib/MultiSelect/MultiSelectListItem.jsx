import React from 'react';
import PropTypes from 'prop-types';
// import { Check } from 'styled-icons/fa-solid/Check';
import styled from 'styled-components';

// TODO (hover background-color) coming from theme? Refactor when decision has been made
const MultiSelectListItemOption = styled('option')`
  padding: 0.625rem 0; /* 10px if base font-size is 16px */
  margin: 1px;
  line-height: 1;
  cursor: pointer;

  :focus {
    outline: 1px solid #5e9ed6;
  }

  :hover {
    background-color: #f0f0f0;
  }

  &.selected {
    display: flex;
    justify-content: space-between;

    &:after {
      content: '';
      display: block;
      width: 3px;
      height: 6px;
      border: solid #000;
      border-width: 0 2px 2px 0;
      transform: rotate(45deg);
    }
  }
`;

// const MultiSelectCheckbox = styled('input')`
//   visibility: hidden;
// `;

// const MultiSelectLisItemLabel = styled('label')`
//   display: flex;
//   justify-content: space-between;
//   padding: 0.625rem 0; /* 10px if base font-size is 16px */
//   margin: 1px;
//   line-height: 1;
//   cursor: pointer;

//   :focus {
//     outline: 1px solid #5e9ed6;
//   }
// `;

// const MultiSelectCheckIcon = styled(Check)`
//   visibility: ${props => (props.checked ? 'visible' : 'hidden')};
//   height: 1rem;
// `;

const MultiSelectListItem = React.forwardRef((props, ref) => {
  const { label, id, selected } = props;

  return (
    <>
      <MultiSelectListItemOption
        {...(selected ? { className: 'multiselect-list-item selected' } : { className: 'multiselect-list-item' })}
        value={id}
        ref={ref}
      >
        {label}
      </MultiSelectListItemOption>
    </>
  );
});

MultiSelectListItem.propTypes = {
  label: PropTypes.string.isRequired,
  name: PropTypes.string,
  id: PropTypes.string.isRequired,
  selected: PropTypes.bool.isRequired
};

export default MultiSelectListItem;
