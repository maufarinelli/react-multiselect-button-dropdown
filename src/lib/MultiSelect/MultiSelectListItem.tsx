import React from 'react';
import styled from 'styled-components';

const MultiSelectListItemLi = styled('li')`
  :hover {
    background-color: #f0f0f0;
  }
`;

const MultiSelectCheckbox = styled('input')`
  visibility: hidden;
`;

const MultiSelectLisItemLabel = styled('label')`
  display: flex;
  justify-content: space-between;
  padding: 0.625rem 0; /* 10px if base font-size is 16px */
  margin: 1px;
  line-height: 1;
  white-space: nowrap;
  cursor: pointer;

  :focus {
    outline: 1px solid #5e9ed6;
  }

  &.is-checked {
    :after {
      content: '';
      display: block;
      width: 5px;
      height: 10px;
      margin-right: 3px;
      border: solid #000;
      border-width: 0 3px 3px 0;
      transform: rotate(45deg);
    }
  }
`;

interface MultiSelectListItemProps {
  label: string;
  id: string | number;
  handleInputChange: (event: React.KeyboardEvent<HTMLLabelElement> | React.ChangeEvent<HTMLInputElement>) => void;
  checked: boolean;
  name?: string;
}

const MultiSelectListItem = React.forwardRef<HTMLLabelElement, MultiSelectListItemProps>((props, ref) => {
  const { label, name, id, handleInputChange, checked } = props;

  return (
    <MultiSelectListItemLi className="multiselect-list-item" role="option">
      <MultiSelectLisItemLabel
        role="checkbox"
        aria-checked={checked}
        tabIndex={0}
        htmlFor={id?.toString()}
        onKeyPress={handleInputChange}
        ref={ref}
        className={checked ? 'multiselect-list-item-label is-checked' : 'multiselect-list-item-label'}
      >
        {label}
        <MultiSelectCheckbox
          className="multiselect-list-item-checkbox"
          type="checkbox"
          name={name}
          id={id?.toString()}
          onChange={handleInputChange}
          checked={checked}
        />
      </MultiSelectLisItemLabel>
    </MultiSelectListItemLi>
  );
});

export default MultiSelectListItem;
