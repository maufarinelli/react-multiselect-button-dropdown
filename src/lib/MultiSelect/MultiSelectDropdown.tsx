import React from 'react';
import styled from 'styled-components';
import MultiSelectBadge from './MultiSelectBadge';

const MultiSelectButton = styled('button')`
  height: 100%;
  padding: 1.25rem; /* 20px if base font-size is 16px */
  font-family: inherit;
  font-size: 1rem;
  cursor: pointer;
  background-color: transparent;
  border: 1px #000 solid;
  border-bottom: 0;
  box-shadow: 0 1px #000;

  :after {
    content: '';
    display: inline-block;
    margin-left: 10px;
    padding: 3px;
    border: solid black;
    border-width: 0 3px 3px 0;
    transform: rotate(45deg);
    -webkit-transform: rotate(45deg);
  }

  &&.is-opened {
    :after {
      transform: rotate(-135deg);
      -webkit-transform: rotate(-135deg);
    }
  }
`;

interface MultiSelectDropdownProps {
  text: string;
  toggleDropdown: () => void;
  quantity: number;
  isOpened: boolean;
}

const MultiSelectDropdown: React.FC<MultiSelectDropdownProps> = ({ text, toggleDropdown, quantity, isOpened }) => (
  <MultiSelectButton
    className={isOpened ? 'multiselect-button-dropdown is-opened' : 'multiselect-button-dropdown'}
    onClick={toggleDropdown}
    aria-haspopup="true"
  >
    <MultiSelectBadge text={text} quantity={quantity} />
  </MultiSelectButton>
);

export default MultiSelectDropdown;
