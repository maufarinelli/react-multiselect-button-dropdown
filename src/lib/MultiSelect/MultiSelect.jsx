import React from 'react';
import PropTypes from 'prop-types';
import {
  MultiSelectWrapper,
  MultiSelectListWrapper,
  MultiSelectList,
  MultiSelectAllButton
} from './MultiSelect.styles';
import MultiSelectDropdown from './MultiSelectDropdown';
import MultiSelectListItem from './MultiSelectListItem';
import MultiSelectFooter from './MultiSelectFooter';

class MultiSelect extends React.PureComponent {
  state = {
    isDropdownOpened: false,
    selectedItems: {}
  };

  constructor(props) {
    super(props);
    this.listItems = [];

    this.keyEvents = {
      up: () => {
        const activeElementIndex = this.listItems.findIndex(item => {
          return item === document.activeElement;
        });
        this.listItems[activeElementIndex - 1] && this.listItems[activeElementIndex - 1].focus();
      },
      down: () => {
        const activeElementIndex = this.listItems.findIndex(item => {
          return item === document.activeElement;
        });
        this.listItems[activeElementIndex + 1] && this.listItems[activeElementIndex + 1].focus();
      },
      home: () => {
        this.listItems[0].focus();
      },
      end: () => {
        this.listItems[this.listItems.length - 1].focus();
      }
    };
  }

  componentDidMount() {
    const { list } = this.props;

    this.setState(
      {
        selectedItems: list.reduce((acc, listItem) => {
          acc[listItem.id] = listItem.selected;
          return acc;
        }, {})
      },
      () => {
        debugger;
        console.log(this.state);
      }
    );
  }

  toggleDropdown = () => {
    const { isDropdownOpened } = this.state;

    this.setState({ isDropdownOpened: !isDropdownOpened }, () => {
      if (this.state.isDropdownOpened) {
        // this.listItems[0].focus();
      }
    });
  };

  handleSelectChange = event => {
    const tag = event.target.tagName;
    const target = tag === 'LABEL' ? event.target.children[0] : event.target;
    // If user used the keyboard to select the label, we need to programatically check the checkbox child.
    // Also needed to work with screen readers
    if (tag === 'LABEL') {
      target.checked = !target.checked;
      // preventing just after toggle checked property in order to prevent the space scroll behavior on a list
      event.preventDefault();
    }
    const { id, checked } = target;

    this.setState(
      prevState => ({
        selectedItems: { ...prevState.selectedItems, [id]: checked }
      }),
      () => {
        this.props.onOptionChanged && this.props.onOptionChanged({ [id]: checked });
      }
    );
  };

  resetSelections = () => {
    this.setState(prevState => ({
      selectedItems: Object.keys(prevState.selectedItems).reduce((acc, listItemName) => {
        acc[listItemName] = false;
        return acc;
      }, {})
    }));
  };

  handleApplyClick = () => {
    const { onSelectionApplied } = this.props;
    const { selectedItems } = this.state;

    onSelectionApplied(selectedItems);
  };

  selectAll = () => {
    this.setState(prevState => ({
      selectedItems: Object.keys(prevState.selectedItems).reduce((acc, listItemId) => {
        acc[listItemId] = true;
        return acc;
      }, {})
    }));
  };

  render() {
    const { isDropdownOpened, selectedItems } = this.state;
    const {
      list,
      dropdownButtonText,
      isRightAligned,
      onSelectionApplied,
      selectAllButtonText,
      resetButtonText,
      applyButtonText
    } = this.props;
    const { handleSelectChange, handleApplyClick, selectAll } = this;
    const selectedItemsValues = Object.keys(selectedItems).filter(itemName => selectedItems[itemName]);
    const selectedItemsQuantity = selectedItemsValues.length;

    return (
      <MultiSelectWrapper className="multiselect-button-dropdown-wrapper">
        <MultiSelectDropdown
          className="multiselect-button-dropdown"
          text={dropdownButtonText}
          quantity={selectedItemsQuantity}
          toggleDropdown={this.toggleDropdown}
          isOpened={isDropdownOpened}
        />
        {isDropdownOpened && (
          <MultiSelectListWrapper
            {...(isRightAligned ? { isRightAligned } : {})}
            className="multiselect-section-wrapper"
          >
            <MultiSelectAllButton className="multiselect-button-select-all" onClick={selectAll}>
              {selectAllButtonText}
            </MultiSelectAllButton>
            <MultiSelectList
              multiple={true}
              hasFooter={!!onSelectionApplied}
              className="multiselect-list"
              keyEvents={{ ...this.keyEvents }}
              value={selectedItemsValues}
              onChange={handleSelectChange}
            >
              {list.map((listItem, index) => {
                const { label, id } = listItem;
                const selected = selectedItems[id];
                const key = `${id}-${index}`;

                return (
                  <MultiSelectListItem
                    className="multiselect-list-item"
                    label={label}
                    key={key}
                    id={id}
                    selected={selected}
                    ref={el => {
                      this.listItems[index] = el;
                      return el;
                    }}
                  />
                );
              })}
            </MultiSelectList>
            {onSelectionApplied && (
              <MultiSelectFooter
                resetSelections={this.resetSelections}
                resetButtonText={resetButtonText}
                applyButtonText={applyButtonText}
                handleApplyClick={handleApplyClick}
              />
            )}
          </MultiSelectListWrapper>
        )}
      </MultiSelectWrapper>
    );
  }
}

MultiSelect.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      name: PropTypes.string,
      selected: PropTypes.bool.isRequired
    })
  ).isRequired,
  dropdownButtonText: PropTypes.string.isRequired,
  isRightAligned: PropTypes.bool,
  selectAllButtonText: PropTypes.string,
  onOptionChanged: PropTypes.func,
  onSelectionApplied: PropTypes.func,
  resetButtonText: PropTypes.string,
  applyButtonText: PropTypes.string
};

MultiSelect.defaultProps = {
  selectAllButtonText: 'Select All'
};

export default MultiSelect;
