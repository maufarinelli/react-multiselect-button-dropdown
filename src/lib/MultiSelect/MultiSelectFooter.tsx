import React from 'react';
import styled from 'styled-components';
import { buttonCSS } from './MultiSelect.styles';

const MultiSelectListFooter = styled('div')`
  display: flex;
  flex-direction: row-reverse;
  padding-top: 1.25rem; /* 20px if base font-size is 16px */
  border-top: 1px #ccc solid;
`;

const MultiSelectApplyButton = styled('button')`
  ${buttonCSS}
`;

interface MultiSelectFooterProps {
  handleApplyClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  applyButtonText?: string;
}

const MultiSelectFooter: React.FC<MultiSelectFooterProps> = ({ applyButtonText = 'Apply', handleApplyClick }) => (
  <MultiSelectListFooter>
    <MultiSelectApplyButton
      className="multiselect-apply-button"
      {...(handleApplyClick && { onClick: handleApplyClick })}
    >
      {applyButtonText}
    </MultiSelectApplyButton>
  </MultiSelectListFooter>
);

export default MultiSelectFooter;
