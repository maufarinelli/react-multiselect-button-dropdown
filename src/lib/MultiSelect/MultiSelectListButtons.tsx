import React from 'react';
import { MultiSelectAllButton, MultiSelectResetButton, MultiSelectListButtonsWrapper } from './MultiSelect.styles';

interface MultiSelectListButtonsProps {
  selectAll: () => void;
  selectAllButtonText?: string;
  resetSelections: () => void;
  resetButtonText?: string;
}

const MultiSelectListButtons: React.FC<MultiSelectListButtonsProps> = ({
  selectAll,
  selectAllButtonText = 'Select All',
  resetSelections,
  resetButtonText = 'Reset',
}) => (
  <MultiSelectListButtonsWrapper>
    <MultiSelectAllButton className="multiselect-button-select-all" onClick={selectAll}>
      {selectAllButtonText}
    </MultiSelectAllButton>
    <MultiSelectResetButton className="multiselect-reset-button" onClick={resetSelections}>
      {resetButtonText}
    </MultiSelectResetButton>
  </MultiSelectListButtonsWrapper>
);

export default MultiSelectListButtons;
