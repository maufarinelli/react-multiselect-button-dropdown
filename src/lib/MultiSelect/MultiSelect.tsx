import React, { useEffect, useRef } from 'react';
import { MultiSelectWrapper, MultiSelectListWrapper, MultiSelectList } from './MultiSelect.styles';
import MultiSelectDropdown from './MultiSelectDropdown';
import MultiSelectListButtons from './MultiSelectListButtons';
import MultiSelectListItem from './MultiSelectListItem';
import MultiSelectFooter from './MultiSelectFooter';
import useMultiSelect from './useMultiSelect';
import { MultiSelectProps } from '../../../types';

const MultiSelect: React.FC<MultiSelectProps> = ({
  list,
  dropdownButtonText,
  isRightAligned,
  selectAllButtonText,
  onOptionChanged,
  onSelectionApplied,
  resetButtonText,
  applyButtonText,
}) => {
  const { isDropdownOpened, checkedItems, toggleDropdown, selectAll, resetSelections, handleInputChange } =
    useMultiSelect({ list });
  const listItems: React.MutableRefObject<Array<HTMLLabelElement | undefined>> = useRef([]);
  const keyEvents = {
    up: () => {
      const activeElementIndex = listItems.current.findIndex((item) => {
        return item === document.activeElement;
      });
      listItems.current[activeElementIndex - 1]?.focus();
    },
    down: () => {
      const activeElementIndex = listItems.current.findIndex((item) => {
        return item === document.activeElement;
      });
      listItems.current[activeElementIndex + 1]?.focus();
    },
    home: () => {
      listItems.current[0]?.focus();
    },
    end: () => {
      listItems.current[listItems.current.length - 1]?.focus();
    },
  };
  const checkedItemsQuantity = Object.keys(checkedItems).filter((itemName) => checkedItems[itemName]).length;

  useEffect(() => {
    if (isDropdownOpened) {
      listItems.current[0]?.focus();
    }
  }, [isDropdownOpened]);

  useEffect(() => {
    onOptionChanged && onOptionChanged(checkedItems);
  }, [onOptionChanged, checkedItems]);

  const handleApplyClick = () => {
    if (onSelectionApplied) {
      onSelectionApplied(checkedItems);
    }
  };

  return (
    <MultiSelectWrapper className="multiselect-button-dropdown-wrapper">
      <MultiSelectDropdown
        text={dropdownButtonText}
        quantity={checkedItemsQuantity}
        toggleDropdown={toggleDropdown}
        isOpened={isDropdownOpened}
      />
      {isDropdownOpened && (
        <MultiSelectListWrapper {...(isRightAligned ? { isRightAligned } : {})} className="multiselect-section-wrapper">
          <MultiSelectListButtons
            selectAll={selectAll}
            selectAllButtonText={selectAllButtonText}
            resetSelections={resetSelections}
            resetButtonText={resetButtonText}
          />
          <MultiSelectList
            role="listbox"
            tag="ul"
            hasFooter={!!onSelectionApplied}
            className="multiselect-list"
            keyEvents={{ ...keyEvents }}
          >
            {list.map((listItem, index) => {
              const { label, id, name } = listItem;
              const checked = checkedItems[id];
              const key = `${id}-${index}`;

              return (
                <MultiSelectListItem
                  label={label}
                  key={key}
                  id={id}
                  name={name}
                  handleInputChange={handleInputChange}
                  checked={checked}
                  ref={(el: HTMLLabelElement) => {
                    listItems.current[index] = el;
                    return el;
                  }}
                />
              );
            })}
          </MultiSelectList>
          {onSelectionApplied && (
            <MultiSelectFooter applyButtonText={applyButtonText} handleApplyClick={handleApplyClick} />
          )}
        </MultiSelectListWrapper>
      )}
    </MultiSelectWrapper>
  );
};

export default MultiSelect;
