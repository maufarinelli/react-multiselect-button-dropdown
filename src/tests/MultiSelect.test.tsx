import React from 'react';
import ReactDOM from 'react-dom/client';
import renderer, { create } from 'react-test-renderer';
import { act } from '@testing-library/react';
import MultiSelect from '../lib/MultiSelect/MultiSelect';
import { MultiSelectProps } from '../../types';

describe('<MultiSelect />', () => {
  let container: HTMLDivElement;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    if (container) {
      document.body.removeChild(container);
    }
  });

  const multiSelectProps: MultiSelectProps = {
    list: [
      {
        label: 'First option',
        name: 'first-option',
        id: 'first-option-1',
        checked: false,
      },
      {
        label: 'Second option',
        name: 'second-option',
        id: 'second-option-2',
        checked: false,
      },
      {
        label: 'Third option',
        name: 'third-option',
        id: 'third-option-3',
        checked: false,
      },
    ],
    onSelectionApplied: (selection) => {
      console.log('Selected : ', selection);
    },
    onOptionChanged: (optionState) => {
      console.log('optionState : ', optionState);
    },
    dropdownButtonText: 'Selected',
    resetButtonText: 'Reset',
    applyButtonText: 'Apply',
  };

  test('should match snapshot - dropdown closed', () => {
    const testRenderer = create(<MultiSelect {...multiSelectProps} />);

    expect(testRenderer.toJSON()).toMatchSnapshot('dropdown closed');
  });

  test('should match snapshot - dropdown opened', () => {
    const testRenderer = create(<MultiSelect {...multiSelectProps} />);
    const testInstance = testRenderer.root;

    renderer.act(() => {
      testInstance.findByProps({ className: 'multiselect-button-dropdown' }).props.onClick();
    });

    expect(testRenderer.toJSON()).toMatchSnapshot('dropdown opened');
  });

  test('should open dropdown when button is clicked', () => {
    const testRenderer = create(<MultiSelect {...multiSelectProps} />);
    const testInstance = testRenderer.root;
    const buttonDropdown = testInstance.findByProps({ className: 'multiselect-button-dropdown' });

    expect(buttonDropdown).toBeDefined();

    renderer.act(() => {
      buttonDropdown.props.onClick();
    });

    const listWrapper = testInstance.findByProps({ className: 'multiselect-section-wrapper' });

    expect(listWrapper).toBeDefined();
  });

  test('should select the first item clicking on its input checkbox', () => {
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    const firstOptionCheckbox: HTMLInputElement | null = firstOptionLabel.querySelector(
      'input.multiselect-list-item-checkbox'
    );

    if (!firstOptionCheckbox) {
      throw new Error('No checkbox found');
    }
    act(() => {
      firstOptionCheckbox.dispatchEvent(new MouseEvent('click'));
    });

    expect(firstOptionCheckbox.checked).toBeTruthy();
  });

  test('should select the first item clicking on its label', () => {
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    const firstOptionCheckbox: HTMLInputElement | null = firstOptionLabel.querySelector(
      'input.multiselect-list-item-checkbox'
    );
    console.log('firstOptionCheckbox : ', firstOptionCheckbox);

    act(() => {
      firstOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    if (!firstOptionCheckbox) {
      throw new Error('No checkbox found');
    }
    expect(firstOptionCheckbox.checked).toBeTruthy();
  });

  test('should unselect the first item', () => {
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    const firstOptionCheckbox: HTMLInputElement | null = firstOptionLabel.querySelector(
      'input.multiselect-list-item-checkbox'
    );

    act(() => {
      firstOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    if (!firstOptionCheckbox) {
      throw new Error('No checkbox found');
    }
    expect(firstOptionCheckbox.checked).toBeTruthy();

    act(() => {
      firstOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    expect(firstOptionCheckbox.checked).toBeFalsy();
  });

  test('should apply selections', () => {
    const spyApply = jest.spyOn(multiSelectProps, 'onSelectionApplied');
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    act(() => {
      firstOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    const applyButton = container.querySelector('button.multiselect-apply-button');
    if (!applyButton) {
      throw new Error('No applyButton found');
    }
    act(() => {
      applyButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyApply).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': false,
      'third-option-3': false,
    });
  });

  test('should badge have the right quantity', () => {
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    act(() => {
      firstOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    const badge = container.querySelector('span.multiselect-badge');
    if (!badge) {
      throw new Error('No badge found');
    }
    expect(badge.textContent).toBe('1');

    const lastOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[2];
    act(() => {
      lastOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    expect(badge.textContent).toBe('2');
  });

  test('should select all items', () => {
    const spyOptionChanged = jest.spyOn(multiSelectProps, 'onOptionChanged');
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const buttonSelectAll = container.querySelector('button.multiselect-button-select-all');
    if (!buttonSelectAll) {
      throw new Error('No buttonSelectAll found');
    }
    act(() => {
      buttonSelectAll.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': true,
      'third-option-3': true,
    });
  });

  test('should reset selection', () => {
    const spyOptionChanged = jest.spyOn(multiSelectProps, 'onOptionChanged');
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const buttonSelectAll = container.querySelector('button.multiselect-button-select-all');
    if (!buttonSelectAll) {
      throw new Error('No buttonSelectAll found');
    }
    act(() => {
      buttonSelectAll.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': true,
      'third-option-3': true,
    });

    act(() => {
      buttonSelectAll.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': false,
      'second-option-2': false,
      'third-option-3': false,
    });
  });

  test('should trigger option changed', () => {
    const spyOptionChanged = jest.spyOn(multiSelectProps, 'onOptionChanged');
    act(() => {
      ReactDOM.createRoot(container).render(<MultiSelect {...multiSelectProps} />);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    if (!buttonDropdown) {
      throw new Error('No buttonDropdown found');
    }
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    act(() => {
      firstOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': false,
      'third-option-3': false,
    });

    const lastOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[2];
    act(() => {
      lastOptionLabel.dispatchEvent(new MouseEvent('click'));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': false,
      'third-option-3': true,
    });
  });
});
