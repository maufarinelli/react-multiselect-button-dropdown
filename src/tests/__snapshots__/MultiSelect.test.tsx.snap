// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelect /> should match snapshot - dropdown closed: dropdown closed 1`] = `
<div
  className="sc-bcXHqe bHDyYL multiselect-button-dropdown-wrapper"
>
  <button
    aria-haspopup="true"
    className="sc-iBYQkv kMZPkI multiselect-button-dropdown"
    onClick={[Function]}
  >
    Selected
     
    <span
      className="sc-gKPRtg giSAku multiselect-badge"
    >
      0
    </span>
  </button>
</div>
`;

exports[`<MultiSelect /> should match snapshot - dropdown opened: dropdown opened 1`] = `
<div
  className="sc-bcXHqe bHDyYL multiselect-button-dropdown-wrapper"
>
  <button
    aria-haspopup="true"
    className="sc-iBYQkv kMZPkI multiselect-button-dropdown is-opened"
    onClick={[Function]}
  >
    Selected
     
    <span
      className="sc-gKPRtg giSAku multiselect-badge"
    >
      0
    </span>
  </button>
  <div
    className="sc-gswNZR knWTbq multiselect-section-wrapper"
  >
    <div
      className="sc-jSUZER cZFivt"
    >
      <button
        className="sc-hLBbgP gNDTIg multiselect-button-select-all"
        onClick={[Function]}
      >
        Select All
      </button>
      <button
        className="sc-eDvSVe rLMxZ multiselect-reset-button"
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
    <ul
      className="sc-dkrFOg kFcxkP multiselect-list"
      onKeyDown={[Function]}
    >
      <li
        className="sc-ftTHYK clYpuu multiselect-list-item"
        role="option"
      >
        <label
          aria-checked={false}
          className="sc-jrcTuL cxAKnR multiselect-list-item-label"
          htmlFor="first-option-1"
          onKeyPress={[Function]}
          role="checkbox"
          tabIndex={0}
        >
          First option
          <input
            checked={false}
            className="sc-pyfCe hMyIYu multiselect-list-item-checkbox"
            id="first-option-1"
            name="first-option"
            onChange={[Function]}
            type="checkbox"
          />
        </label>
      </li>
      <li
        className="sc-ftTHYK clYpuu multiselect-list-item"
        role="option"
      >
        <label
          aria-checked={false}
          className="sc-jrcTuL cxAKnR multiselect-list-item-label"
          htmlFor="second-option-2"
          onKeyPress={[Function]}
          role="checkbox"
          tabIndex={0}
        >
          Second option
          <input
            checked={false}
            className="sc-pyfCe hMyIYu multiselect-list-item-checkbox"
            id="second-option-2"
            name="second-option"
            onChange={[Function]}
            type="checkbox"
          />
        </label>
      </li>
      <li
        className="sc-ftTHYK clYpuu multiselect-list-item"
        role="option"
      >
        <label
          aria-checked={false}
          className="sc-jrcTuL cxAKnR multiselect-list-item-label"
          htmlFor="third-option-3"
          onKeyPress={[Function]}
          role="checkbox"
          tabIndex={0}
        >
          Third option
          <input
            checked={false}
            className="sc-pyfCe hMyIYu multiselect-list-item-checkbox"
            id="third-option-3"
            name="third-option"
            onChange={[Function]}
            type="checkbox"
          />
        </label>
      </li>
    </ul>
    <div
      className="sc-kDvujY ewaAmn"
    >
      <button
        className="sc-ipEyDJ dNOPPb multiselect-apply-button"
        onClick={[Function]}
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
